llvm/lib/MC/MCPseudoProbe.cpp

```diff
@@ -443,17 +446,24 @@ bool MCPseudoProbeDecoder::buildAddress2ProbeMap(
   // If the incoming node is null, all its children nodes should be disgarded.
   if (Cur) {
     // Switch/add to a new tree node(inlinee)
-    Cur = Cur->getOrAddNode(std::make_tuple(Guid, Index));
-    Cur->Guid = Guid;
+    Cur->Children[CurChild] = MCDecodedPseudoProbeInlineTree(Guid, Index, Cur);
+    Cur = &Cur->Children[CurChild];
     if (IsTopLevelFunc && !EncodingIsAddrBased) {
       if (auto V = FuncStartAddrs.lookup(Guid))
         LastAddr = V;
     }
   }
+  // Advance CurChild for non-skipped top-level functions and unconditionally
+  // for inlined functions.
+  if (IsTopLevelFunc)
+    CurChild += !!Cur;
```

Comment:
Is it possible to move the index advancing things outside of the function(to the caller)?

For non-top-level funcs, just like the typical iteration: 
```
for (uint32_t I = 0; ...; I++) 
  ....
```
which would be better for the readability? also then we don't need to pass it by reference.

Reply:
It was initially this way, but I had to change to advancing CurChildIndex from within buildAddress2ProbeMap. The problem is if GuidFilter is in place, we will only allocate enough top-level entries for filtered functions. Therefore we can't advance CurChildIndex from top-level buildAddress2ProbeMap invocation (GUID is not yet parsed).

We can advance non-top level child index though.

Let me refactor this a bit so it's easier to follow.